pythoncode/bleDongle.py

```diff
@@ -0,0 +1,325 @@
+# c l s B l e I n t e r f a c e
+#---------------------------------------------------------------------------
+# Input:    UseBluetooth
+#               If False, it will be a 'dummy' class, just providing the data-
+#               structure so that the caller can refer without checking the
+#               directory on every location.
+#
+#           host and port are variable, but note that Node.js must be
+#               redefined as well in that case.
+#
+# Function  The Bluetooth interface object
+#           Providing elementary functions to Send/receive data
+#
+#---------------------------------------------------------------------------
+class clsBleInterface():
+    def __init__(self, clv, host = 'localhost', port = 9999):
+        self.OK        = False
+        self.host      = host
+        self.port      = port
+        self.interface = None
+        self.jsondata  = None
+        if UseBluetooth and clv.ble:
+            self.Message   = ", Bluetooth interface available"
+            #---------------------------------------------------------------
+            # register self.Close() to make sure the BLE server is stopped
+            #   ON program termination
+            # Note that __del__ is called too late to be able to close.
+            #---------------------------------------------------------------
+            atexit.register(self.Close)
+        else:
+            self.Message   = "" # Bluetooth disabled or not used
+
+    def __del__(self):
+        pass
+
+    if UseBluetooth:
+        #-------------------------------------------------------------------
+        # O p e n
+        #-------------------------------------------------------------------
+        # input     none
+        #
+        # function  The Bluetooth interface is opened
+        #
+        # returns   self.interface
+        #-------------------------------------------------------------------
+        def Open(self):
+            self.OK = False
+            self.Message = ", Bluetooth interface cannot be opened"
+            if debug.on(debug.Ble): logfile.Write ("BleInterface.Open() ...")
+
+            if self.interface:
+                if debug.on(debug.Ble): logfile.Write('... already open')
+            else:
+                #-----------------------------------------------------------
+                # Create interface as sub-process
+                #-----------------------------------------------------------
+                command = [
+                    "node",
+                    "server.js"
+                ]
+                directory = Path.cwd().parent / "node"
+                if debug.on(debug.Ble): logfile.Write("... Popen(%s,%s)" % (directory, command) )
+                try:
+                    if debug.on(debug.Any):
+                        self.interface = subprocess.Popen(command, cwd=directory, stdout=logfile.fLogfile, stderr=logfile.fLogfile)
```

Comment:
Currently the default logging in the BLE server is very minimal. If you want to debug issues, setting the `DEBUG` environment variable to `*` may help.

```python
env = os.environment
env['DEBUG'] = '*'
self.interface = subprocess.Popen(command, cwd=directory, stdout=logfile.fLogfile, stderr=logfile.fLogfile)
```